components/constellation/constellation.rs

```diff
@@ -5388,29 +5428,26 @@ where
             })
     }
 
-    /// Re-send the frame tree to the compositor.
-    fn update_frame_tree_if_active(
+    /// Send the frame tree for the given browser to the compositor.
+    fn update_frame_tree_if_focused(
         &mut self,
         top_level_browsing_context_id: TopLevelBrowsingContextId,
     ) {
-        // Only send the frame tree if it's the active one or if no frame tree
-        // has been sent yet.
-        if self.active_browser_id.is_none() ||
-            Some(top_level_browsing_context_id) == self.active_browser_id
-        {
-            self.send_frame_tree(top_level_browsing_context_id);
+        // Only send the frame tree if the given browser is focused, or if no
+        // browser is currently focused, make it focused.
+        if let Some(focused_browser_id) = self.browsers.focused_browser().map(|(id, _)| id) {
+            if top_level_browsing_context_id != focused_browser_id {
+                return;
+            }
+        } else {
+            self.browsers.focus(top_level_browsing_context_id);
```

Comment:
Is this what we'd like to do? For instance, imagine a web view that is integrated into some other UI system (this actually happens a lot). This will cause the web view to focus meaning an input element might start to appear active -- in addition to whatever external UI element has focus. Shouldn't the embedder completely control when the web view has focus?

Reply:
yeah good point, i think i did this to preserve existing behaviour since i didn’t understand the session history and frame tree stuff well enough, but i agree. fixed in 2afa67966ac13eb312cbf0fa616cbfe7c3790f4f